### PR TITLE
[Testing] Improve testing capabilities for Web

### PR DESCRIPTION
**Pull request by RobertBroersma**
_related to #502_

This PR is for writing tests in the `web` side!

# Routes

Normally when using `@reach/router` or `react-router` I wouldn't mock my `<Router />` component, I would just wrap the render in it, but `@redwoodjs/router` makes that difficult because

- `<Router />` doesn't accept children.
- `<Router />` does some initial magic with pages that you _probably_ don't want in your test.

I did 2 things to "solve" these issues:

## 1. Import the `Routes.js` file _from_ the application _to_ the Redwood library.

This line in `jest.config.web.js` does exactly that. `<rootDir>` is the application root.

```javascript
import Routes from '~/web/Routes': '<rootDir>/src/Routes',
```

By rendering the `<Routes />` component _before_ the subject, all routes should be calculated before they are used in the subject.

## 2. Mock `<Router />`

I can't quite figure out why, but I had some difficulties with `page-loader.js` when not mocking `<Router />`. In order for any of this to work I put this mock in `__mocks/@redwoodjs/routes.js`:

```javascript
export * from "@redwoodjs/router";

const routes = {};

export const Router = ({ children }) => {
  for (let route of React.Children.toArray(children)) {
    const { name } = route.props;
    routes[name] = jest.fn(() => name);
  }

  return <></>;
};

export { routes };
```

This still enables using `routes.nameOfRoute` in your tests, but mocks out the rest, meaning nothing of `<Routes />` or `<Router />` will also show up in your tests!

I'm not sure where this mock should live. I think keeping all mocks on the application side could be the best, so there's no hidden differences in application and test going on for the user. It might also be difficult to have this mock live inside `node_modules`, Jest doesn't deal with that.

One thing we _could_ do is keep it in `node_modules` and `import` it inside the application mock.

# Apollo MockProvider

Before we come up with some fancy tool that knows about your backend and can assist you in your tests, I think the best would be to just support mocking out the backend. This is pretty straightforward to do with Apollo's `MockProvider`!

# Implementation

I've created a small repo that demonstrates the usage of this PR: https://github.com/RobertBroersma/redwood-testing

I went through about half of the tutorial to get some Cells. `web/src/pages/BlogPostPage/BlogPostPage.test.js` contains all the stuff I mentioned!

# Conclusion

It works, but it's still WIP! As always happy to receive any feedback.

One bullet would be where to keep all these files? I suppose some of that jest config could be moved to testing package? Could the testing package be small enough to merge into `core`? Would it grow in the future?

PS. Please ignore the babel cell plugin stuff. It's in a separate PR #512 . I need that for Jest to work with Cells. Feel free to ignore this until #512 is merged!

### CODE REVIEW

File: packages/testing/src/render.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      </>
+    </MockedProvider>
+  )
+}
+
+export const customRender = (ui, options = {}) =>
+  render(ui, {
+    wrapper: (props) => <AllTheProviders mocks={options.mocks} {...props} />,
+    ...options,
+  })
```
Comment (cball) [inline packages/testing/src/render.js:22]: I'm curious if we can improve the API here?

Currently a test would use:

```js
render(<BlogPostPage id="id-123" />, { mocks })
```

In a more complex situation, it might eventually look like this:

```js
render(<BlogPostPage id="id-123" />, { mocks, useTypeName: true, resolvers: localResolvers,  defaultOptions: apolloOptions })
```

While these options are less common, it may be good to still expose them (especially `useTypeName` for reasons I mentioned above). It feels like in doing so we will start overloading the `render` function too much.

What are your thoughts on exporting an additional top level component from this file that controls mocking, and exporting the custom render separately without the mocks? This would have the additional benefit of allowing unmocked Apollo queries should we need to, as well as handle additional [Apollo options](https://www.apollographql.com/docs/react/api/react-testing/#mockedprovider) more easily.

It might look something like this:

```ts
interface MockedTestStateProps {
  children: ReactNode;
  mockOptions: MockedProviderProps;
}

export const WithMockedTestState = ({ children, mockOptions = {}}: MockedTestStateProps) => {
  <MockedProvider {...mockOptions}>
    {children}
  </MockedProvider>
}
```

The example above would be written as:

```js
render(
  <WithMockedTestState mocks={mocks}>
    <BlogPostPage id="id-123" />
  </WithMockedTestState>
)
```

And for the complex case:

```js
render(
  <WithMockedTestState mocks={mocks} resolvers={localResolvers} defaultOptions={apolloOptions}>
    <BlogPostPage id="id-123" />
  </WithMockedTestState>
)
```

When Apollo mocking is either not desired, or the user is testing a component that doesn't use Apollo, it could still be written as:

```js
render(
  <BlogPostPage id="id-123" />
)
```

As a future point, we may want to consider a way to pass additional `Providers` to the custom render (like theme or other context providers) but for now, I think it's more than fine they are passed explicitly.

**Comment by cball** [comment]
@RobertBroersma awesome work! re: your point on `2. Mock <Router />`, that's something you'd typically have to do when using `next.js` as well.

I like your idea of automatically handling the mock in Redwood and maybe in the future the mocked `jest.fn` could be exposed so that users could assert that it was called with the correct parameters.

File: packages/testing/src/render.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import React from 'react'
+import { render } from '@testing-library/react'
+import { MockedProvider } from '@apollo/react-testing'
+
+import Routes from '~/web/Routes'
+
+const AllTheProviders = ({ mocks, children }) => {
+  return (
+    <MockedProvider mocks={mocks} addTypename={false}>
```
[the remark this answers is not in the document]
Comment (RobertBroersma) [reply packages/testing/src/render.js:9]: I added that part because the Apollo docs said so: https://www.apollographql.com/docs/react/development-testing/testing/

I'm not exactly sure why tho, so if you think it's needed I'm happy to make it overridable!

File: packages/testing/src/render.js (hunk 1 above)
Comment (RobertBroersma) [reply packages/testing/src/render.js:22]: I see your point. I haven't used Apollo Mocks before at all so it's good to get the perspective from someone who has!

I guess my idea was to save people from having to wrap the `MockProvider` in every test where the mock part _isn't_ complicated, which I can imagine are most tests.

I like the idea of splitting them up so it's more flexible tho. But in that case I'm wondering what is the additional benefit of the component we put in between as opposed to just importing the entire `MockProvider` in your app. Unless of course we want to add other stuff later! In that case I think it's good to just re-export `MockProvider` directly from apollo. Or maybe it's an idea to export 2 renders; one with and one without the mocks. But that might be messy/weird? What do you think? 

As for adding additional Providers, this is already possible! You can use the `render` function exposed by this PR, and just go:
```javascript
render(
  <Provider>
    <Subject />
  </Provider>
)
```

You could even extend the render function again in your application if you'd like!

**Comment by RobertBroersma** [comment]
@cball Yeah I've been checking out Gatsby docs as well and they recommend mocking out Gatsby entirely 😄 

Thanks again for the feedback. Like I said it's very helpful to get some perspective from someone with actual experience using this stuff!

File: packages/testing/src/render.js (hunk 1 above)
Comment (cball) [reply packages/testing/src/render.js:22]: I'll let someone from the Core Team chime in on this. This is really an API design consideration since both are functionally equivalent.

To me, explicitly adding the wrapper component for mocks is clearer and more explicit than adding additional options to the React Testing Library render function. Of course, it's entirely possible that I'm the outlier in that thought. 

```js
render(
  <WithMockedTestState mocks={mocks}>
    <BlogPostPage id="id-123" />
  </WithMockedTestState>
)
```
vs

```js
render(
  <BlogPostPage id="id-123" />,
  { mocks }
)
```

File: packages/core/config/fileMock.js [hunk 3]
```diff
@@ -0,0 +1 @@
+module.exports = 'fileMock'
```
Comment (peterp) [inline packages/core/config/fileMock.js:1]: Do you think this file could be part of `@redwoodjs/testing`

Comment (peterp) [inline packages/core/config/fileMock.js:1]:
```suggestion
/**
 * configure Jest to gracefully handle asset files such as stylesheets and images.
 * Usually, these files aren't particularly useful in tests so we can safely mock them out.
 * https://jestjs.io/docs/en/webpack#handling-static-assets
 */
module.exports = 'fileMock'
```

File: packages/core/config/jest.config.web.js [hunk 4]
```diff
@@ -2,11 +2,29 @@
...
+
+const redwoodPaths = getPaths()
+
 module.exports = {
   resolver: 'jest-directory-named-resolver',
   rootDir: process.cwd(),
   globals: {
     __REDWOOD__API_PROXY_PATH: '/',
   },
   setupFilesAfterEnv: [path.resolve(__dirname, './jest.setup.web.js')],
```
Comment (peterp) [inline packages/core/config/jest.config.web.js:15]: I know you didn't change this, but I wanted to make a note for myself to change this.
```suggestion 
  // TODO: use getPaths().web.base instead.
  setupFilesAfterEnv: [path.resolve(__dirname, './jest.setup.web.js')],
```

File: packages/core/config/jest.config.web.js [hunk 5]
```diff
@@ -2,11 +2,29 @@
...
+  moduleNameMapper: {
+    '^react$': path.resolve(redwoodPaths.base, 'node_modules', 'react'),
+    '^react-dom$': path.resolve(redwoodPaths.base, 'node_modules', 'react-dom'),
+    '^@apollo/react-common': path.resolve(
+      redwoodPaths.base,
+      'node_modules',
+      '@apollo/react-common'
+    ),
```
Comment (peterp) [inline packages/core/config/jest.config.web.js:23]: Could you explain what these do in a comment?

File: packages/testing/src/render.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import React from 'react'
+import { render } from '@testing-library/react'
+import { MockedProvider } from '@apollo/react-testing'
+
+import Routes from '~/web/Routes'
```
Comment (peterp) [inline packages/testing/src/render.js:5]: I think we might be able to rely on a dynamic import here, and using `getPaths().web.routes`. If not, maybe we could configure the configuration to use `getPaths` because I want to have a single handle on the location of this file.

File: packages/testing/src/index.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export * from '@testing-library/react'
+
+export { customRender as render } from './render'
```
Comment (peterp) [inline packages/testing/src/index.js:3]:
```suggestion
// https://testing-library.com/docs/react-testing-library/setup#custom-render
export { customRender as render } from './render'
```

File: packages/testing/src/render.js (hunk 1 above)
Comment (peterp) [reply packages/testing/src/render.js:22]: I think both approaches have merit: 

1. The `render(Component, { mocks: [] })` is nice because it's less typing, but it may be unexpected. 
2. Using a component is more explicit, but it isn't very different to importing `MockProvider` and wrapping it yourself.

I don't think we have a very compelling integration path for mocking graphql response data yet, but we'll get there. I think we should stick with the `render` argument, but we could name it something else:
```js
render(..., { sdl: { mocks: {}, resolvers: {}, ...etc } } )
```
Maybe we could conditionally render the `MockProvider` if `sdl` is present?

Some ideas for the future:
1. Have an single place to put mock data that works across all the layers of the `db -> sdl -> resolvers -> services.`
2. Figure out a way to generate mock data based off "field names". As an example we could easily generate somewhat realistic data for the following: name, surname, address, title, slug, email, and for the fields that cannot be determined we could figure something out.

**Comment by peterp** [comment]
@cball Thank you for chiming in here too! You made me consider a bunch of things that wasn't really in my mind. Much appreciated!

File: packages/core/config/fileMock.js (hunk 3 above)
Comment (RobertBroersma) [reply packages/core/config/fileMock.js:1]: I think it could. Do you think it would be a good idea to move other testing stuff like Jest config and the setup files to that `@redwoodjs/testing` as well? And have the testing package sort of dogfood the rest of the monorepo?

File: packages/testing/src/render.js (hunk 1 above)
Comment (RobertBroersma) [reply packages/testing/src/render.js:22]: It took me a bit, but I think I understand where you are going with the `sdl` and mocking the full api or just `resolvers` or even other layers.

I do think it's quite a big API decision like @cball mentioned. 

How would you feel about taking the `MockedProvider` out of the custom `render` method for now, writing a small recipe/guide/doc for current users on how to import `MockedProvider` and then we can discuss a bit more on what the API should look like...

OR

We leave it in for now, perhaps in the format you suggested, we see what people think and change it later... Redwood is in early stages after all.

**Comment by RobertBroersma** [comment]
@cball We went ahead and removed `MockedProvider` from the default `render` export until we come up with an API that's certain to satisfy all!

**Comment by RobertBroersma** [comment]
(I'm not entirely sure what's wrong with my build checks)

File: packages/core/config/fileMock.js (hunk 3 above)
Comment (peterp) [reply packages/core/config/fileMock.js:6]: I would like to keep configuration in core. In my mind the "core" package is what a user's requires to develop and build a redwood project and the configurations are part of that.

The file mock feels like something that should belong to test, whereas this configuration is something that's using "@redwoodjs/testing" - that distinction may seem a bit loosey-goosey and I'll try to clarify it with better language.

**Comment by thedavidprice** [comment]
@RobertBroersma this is looking so. good. Just 🔥Some misc topics for discussion as you are nearing the end!

## Documentation
In general, we're trying to create a distinction between docs for 1) contributing to Redwood and 2) developing with Redwood. The former is the content in package READMEs, and the latter is the Guides section of redwoodjs.com and lives in the related repo docs/ directory. Here's my current (very much draft) attempt to add some guidelines #332 

For the docs you've started here, can we:
- move the "developing" related content to a new Issue on redwoodjs.com repo
- use the README stub structure (from 332) to create the README here?

I'll help with any/all of this. Also looping in @jtoar Just say the word!

## Jest --config extensibility
I want to keep this _future_ requirement on the radar (note: it's definitely _not_ a "must include" in this PR). 
> if additional config is needed, capability for simple config extension managed from either App root or specific to App sides (e.g. web/, api/, etc.).

The current design of the `yarn rw test` command along with jest config in the core package makes it non-intuitive how to extend/add config if needed.

Robert, if you already have an idea about how this could work, near-term I could help give some guidance via the Testing Docs (if possible). Otherwise, we'll get to this down the road but let's not lose track of it.

## Updating the RW Generator `.test.js` templates
Our goal for Redwood's testing support is that individuals will be able to go through the Tutorial and end up with _passing tests_ along with the app they created.

Our current test templates will need to be updated to match the work you've done here. Just point us in the right direction when you're ready and we'll make sure it happens!

---

Thanks again!

**Comment by RobertBroersma** [comment]
@thedavidprice Thanks for the input!

> For the docs you've started here, can we:
>
> move the "developing" related content to a new Issue on redwoodjs.com repo
> use the README stub structure (from 332) to create the README here?

:heavy_check_mark: Will do! I'll tag you and @jtoar when I do to make sure it's all in the right place and the language is good.

> ## Jest --config extensibility

Important to keep on the radar indeed! Atm you could extend the config by directly importing the jest config from core and merging it with your local jest config, however this is 
1. Currently undocumented
2. Not optimal...? I'm thinking along the way of exposing the Redwood jest config(s) as a function that you could call. Something like:
```javascript
import { createJestConfig } from '@redwoodjs/core' (or @redwoodjs/testing)

export default createJestConfig({
 ...myCustomStuff,
})
```
(As there is no `extend` feature in Jest like in Babel/Eslint)

Let's create a separate GitHub issue to continue discussing this? :rocket: 

About targeting different sides; I thought it could be good to use Jest's [`projects`](Jest --config extensibility) config instead of running all sides in the command, so we offload more stuff to the Jest side, and new sides with their own jest.configs are automatically picked up!

One more thing I'd also like to do in the future is add https://github.com/jest-community/jest-watch-typeahead to the base config. This of course requires `yarn rw test` to work with watch!

> ## Updating the RW Generator `.test.js` templates

I think we for sure should Include the `__mock__/@redwoodjs/router` mock I haphazardly documented in the README on this PR by default. This would be required to get the tests with `routes.x()` in them to work

Another thing _I think_ we should do is change the generated tests for Cells (Partially like in the README on this PR + Some extra tests for `Failure` and `Empty`). This is however up for debate, as merging this won't break the current test setup. The new suggested setup would just push the tests more from unit tests to integration tests (if we'd have to slap a label on it). 

Let me know if you'd like me to create some of those new issues I mentioned.

---

Cheers!

**Comment by thedavidprice** [comment]
**Documentation** --> perfect. And don't hesitate to offload as needed!

**Jest --config** --> Done #564  Ok by me to start with the "documenting" option. Maybe add to current round of docs?

**Templates** --> agreed on both counts for Pages and Cells (we could definitely use better template code for Cells).  I'll assume you will take lead, but I’m happy to help if you let me know !

File: packages/testing/src/render.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import React from 'react'
+import { getPaths } from '@redwoodjs/internal'
+import { render } from '@testing-library/react'
+
+const redwoodPaths = getPaths()
+
+const Routes = require(`${redwoodPaths.web.src}/Routes.js`).default
```
Comment (peterp) [inline packages/testing/src/render.js:7]:
```suggestion
const Routes = require(`${redwoodPaths.web.routes`).default
```

**Comment by RobertBroersma** [comment]
Thanks @peterp !

@thedavidprice I added a _very concise_ implementation of the Readme.md stub you linked. Feel free to add more details if you have the time! I wasn't sure who to put for package lead. I put myself for now, but maybe that should be someone from the core team? :smile: 

I have some small changes to the mock instructions since last time. I will open up a PR on redwoodjs.com repo asap :)

**Comment by RobertBroersma** [comment]
> Hey @RobertBroersma I'm trying to get this to run in `example-blog`, but it doesn't appear to be able to import the pages correctly.
> 
> Did you have an example app where you got this working?

My bad! I had some small changes to the example, forgot to push them. If you pull on master it should all work.

**Comment by RobertBroersma** [comment]
@peterp I see now that you mentioned `example-blog`, not my `redwood-testing` repo! Did you take a look at that one for the web/`__mocks__/@redwoodjs/router.js` mock file? That is a requirement to get the `routes.route()` functions to work. I will open a doc PR and a PR for generating thing file asap!
